File: Python/BOJ/해시/20166.py
//
//  20166.py
//  SwiftAlgorithm
//
//  Created by 유정주 on 2023/06/13.
//

import sys; readLine = sys.stdin.readline
from collections import deque

dx = [-1, -1, 0, 1, 1, 1, 0, -1]
dy = [0, 1, 1, 1, 0, -1, -1, -1]

def bfs(x, y):
    global dict
    q = deque([(x, y, board[x][y], 1)])
    
    while q:
        x, y, word, depth = q.popleft()

        if depth > 5:
            continue
        
        dict[word] = dict.get(word, 0) + 1

        for i in range(8):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx < 0:
                nx += n
            if ny < 0:
                ny += m
            if nx >= n:
                nx -= n
            if ny >= m:
                ny -= m

            q.append((nx, ny, word + board[nx][ny], depth + 1))

n, m, k = map(int, readLine().split())
board = [list(readLine().rstrip()) for _ in range(n)]
arr = [readLine().rstrip() for _ in range(k)]

dict = {}

for i in range(n):
    for j in range(m):
        bfs(i, j)

for string in arr:
    print(dict.get(string, 0))

File: Swift/BOJ/20000/20166.swift
//
//  20166.swift
//  Algorithm
//
//  Created by 유정주 on 2023/06/13.
//

import Foundation

let dx = [-1, -1, 0, 1, 1, 1, 0, -1]
let dy = [0, 1, 1, 1, 0, -1, -1, -1]

func dfs(_ x: Int, _ y: Int, _ current: String, _ depth: Int) {
    if depth > 5 {
        return
    }
    
    dict[current, default: 0] += 1
    
    for i in 0..<8 {
        var nx = x + dx[i]
        var ny = y + dy[i]
        
        if nx < 0 { nx += n }
        if nx >= n { nx -= n }
        
        if ny < 0 { ny += m }
        if ny >= m { ny -= m }
        
        dfs(nx, ny, current + board[nx][ny], depth + 1)
    }
}

let nmk = readLine()!.split { $0 == " " }.map { Int($0)! }
let n = nmk[0], m = nmk[1], k = nmk[2]

var board: [[String]] = []
for _ in 0..<n {
    let input = Array(readLine()!).map { String($0) }
    board.append(input)
}

var arr: [String] = []
for _ in 0..<k {
    let input = readLine()!
    arr.append(input)
}

var dict: [String: Int] = [:]

for i in 0..<n {
    for j in 0..<m {
        dfs(i, j, board[i][j], 1)
    }
}

for string in arr {
    print(dict[string] ?? 0)
}
